modules/phase_field/src/userobjects/PolycrystalVoronoi.C

```diff
@@ -166,6 +209,59 @@ PolycrystalVoronoi::precomputeGrainStructure()
         _centerpoints[grain](2) = _bottom_left(2) + _range(2) * 0.5;
     }
   }
+
+  // Build a KDTree that is used to speedup point search
+  buildSearchTree();
+}
+
+void
+PolycrystalVoronoi::buildSearchTree()
+{
+  if (!_use_kdtree)
+    return;
+
+  auto midplane = _bottom_left + _range / 2.0;
+  dof_id_type lgrianid = 0;
+  _grain_gtl_ids.clear();
+  _grain_gtl_ids.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  _new_points.clear();
+  // Use more memory. Factor is 2^dim
+  _new_points.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  // Domain will be extended along the periodic directions
+  // For each direction, a half domain is constructed
+  std::vector<std::vector<Real>> xyzs(3);
+  for (auto & point : _centerpoints)
+  {
+    // Cear up
+    for (unsigned int i = 0; i < 3; i++)
+      xyzs[i].clear();
+
+    // Have at least the original point
+    for (unsigned int i = 0; i < 3; i++)
+      xyzs[i].push_back(point(i));
+
+    // Add new coords when there exists periodic boundary conditions
+    // We extend half domain
+    for (unsigned int i = 0; i < _mesh.dimension(); i++)
+      if (_mesh.isTranslatedPeriodic(_vars[0]->number(), i))
+        xyzs[i].push_back(point(i) <= midplane(i) ? point(i) + _range(i) : point(i) - _range(i));
+
+    // Construct all combinations
+    for (auto x : xyzs[0])
+      for (auto y : xyzs[1])
+        for (auto z : xyzs[2])
+        {
+          _new_points.push_back(Point(x, y, z));
+          _grain_gtl_ids.push_back(lgrianid);
+        }
+
+    lgrianid++;
+  }
+
+  // Build a KDTree that is used to speedup point search
+  // We should not destroy _new_points after the tree is contributed
+  // KDTree use a reference to "_new_points"
+  _kd_tree = libmesh_make_unique<KDTree>(_new_points, _mesh.getMaxLeafSize());
```

Comment:
I believe we officially have moved to C++14. You don't have to use the libMesh wrapper for unique anymore.

Reply:
Not sure if I should believe this statement. @lindsayad had told me several times before: we supported C++14, but civet tried to beat back if tried to use `std::make_unique`

@lindsayad So we finally supported   C++14 now?

Reply:
Look at @permcody back to his reviewing ways!! We have not made C++14 required yet. See https://github.com/idaholab/moose/pull/17581